compiler/rustc_const_eval/src/interpret/validity.rs

```diff
@@ -489,24 +489,25 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
                             }
                             None => {}
                         }
-                        // Return alloc mutability
-                        if let DefKind::Static { mt, nested } = self.ecx.tcx.def_kind(did) {
-                            if nested {
-                                mt
-                            } else if !self
-                                .ecx
-                                .tcx
-                                .type_of(did)
-                                .no_bound_vars()
-                                .expect("statics should not have generic parameters")
-                                .is_freeze(*self.ecx.tcx, ty::ParamEnv::reveal_all())
+                        // Return alloc mutability. For "root" statics we look at the type to account for interior
+                        // mutability; for nested statics we have no type and directly use the annotated mutability.
+                        match self.ecx.tcx.def_kind(did) {
+                            DefKind::Static { mt: Mutability::Mut, .. } => Mutability::Mut,
+                            DefKind::Static { mt: Mutability::Not, nested: true } => {
+                                Mutability::Not
+                            }
+                            DefKind::Static { mt: Mutability::Not, nested: false }
+                                if !self
+                                    .ecx
+                                    .tcx
+                                    .type_of(did)
+                                    .no_bound_vars()
+                                    .expect("statics should not have generic parameters")
+                                    .is_freeze(*self.ecx.tcx, ty::ParamEnv::reveal_all()) =>
                             {
                                 Mutability::Mut
-                            } else {
-                                mt
                             }
-                        } else {
-                            Mutability::Not
+                            _ => Mutability::Not,
```

Comment:
The nice thing about the old code structure: it was clear that `DefKind::Static` was fully dealt with. With this match it's less obvious whether any `DefKind` static cases fall through to this `_` arm.